src/v/raft/consensus.cc

```diff
@@ -2147,109 +2158,108 @@ consensus::install_snapshot(install_snapshot_request&& r) {
           // least last snapshot index
           _flushed_offset = std::max(_last_snapshot_index, _flushed_offset);
       });
 }
 
-ss::future<> consensus::do_hydrate_snapshot(storage::snapshot_reader& reader) {
-    return reader.read_metadata()
-      .then([this](iobuf buf) {
-          auto parser = iobuf_parser(std::move(buf));
-          auto metadata = reflection::adl<snapshot_metadata>{}.from(parser);
-          vassert(
-            metadata.last_included_index >= _last_snapshot_index,
-            "Tried to load stale snapshot. Loaded snapshot last "
-            "index {}, current snapshot last index {}",
-            metadata.last_included_index,
-            _last_snapshot_index);
+ss::future<std::optional<snapshot_metadata>>
+consensus::read_snapshot_metadata() {
+    std::optional<raft::snapshot_metadata> metadata;
+    auto snapshot_reader = co_await _snapshot_mgr.open_snapshot();
+    if (!snapshot_reader.has_value()) {
+        co_return std::nullopt;
+    }
+    std::exception_ptr eptr;
+    try {
+        auto buf = co_await snapshot_reader->read_metadata();
+        auto parser = iobuf_parser(std::move(buf));
+        metadata = reflection::adl<raft::snapshot_metadata>{}.from(parser);
+    } catch (...) {
+        eptr = std::current_exception();
+    }
+    co_await snapshot_reader->close();
+    if (eptr) {
+        rethrow_exception(eptr);
+    }
+    co_return metadata;
+}
 
-          vlog(
-            _ctxlog.info,
-            "hydrating snapshot with last included index: {}",
-            metadata.last_included_index);
-
-          _last_snapshot_index = metadata.last_included_index;
-          _last_snapshot_term = metadata.last_included_term;
-
-          // TODO: add applying snapshot content to state machine
-          auto prev_commit_index = _commit_index;
-          _commit_index = std::max(_last_snapshot_index, _commit_index);
-          maybe_update_last_visible_index(_commit_index);
-          if (prev_commit_index != _commit_index) {
-              _commit_index_updated.broadcast();
-              _event_manager.notify_commit_index();
-          }
+void consensus::load_from_metadata(const raft::snapshot_metadata& metadata) {
+    vlog(
+      _ctxlog.info,
+      "hydrating snapshot with last included index: {}",
+      metadata.last_included_index);
 
-          update_follower_stats(metadata.latest_configuration);
-          return _configuration_manager
-            .add(_last_snapshot_index, std::move(metadata.latest_configuration))
-            .then([this, delta = metadata.log_start_delta]() mutable {
-                _probe->configuration_update();
+    _last_snapshot_index = metadata.last_included_index;
+    _last_snapshot_term = metadata.last_included_term;
 
-                if (delta < offset_translator_delta(0)) {
-                    delta = offset_translator_delta(
-                      _configuration_manager.offset_delta(
-                        _last_snapshot_index));
```

Comment:
would this diff be more readable if new code were added at the end of the consensus.cc file or are these functions that are all +/- interleaved really changing this much?

Reply:
I think I can pull this refactoring out to a separate commit and maybe that'd be clearer

Reply:
Done